content/blog/anarchos/2024-04-09_an_odissey_to_port_compcert.md

```diff
@@ -0,0 +1,94 @@
+
+### The normal part
+
+Try to install `Coq` : 
+`opam install coq` --> The coq package relies on the (ocaml) package `lablgtk3`. This one is used in `coqide` an IDE for Coq.
+The critical component used there is `gtksourceview3`, used to render Coq source files with syntax coloring. 
+
+### The journey begins...
+For months or even more than a year, i could not get past this lablgtk3 package, cause *Haiku* only gets `gtksourceview-4` in `haikuports/x11-libs/gtksourceview`.
+So i tried to trick lablgtk3 to compile with gtksourceview-3 files symlinked to gtksourceview-4 files, and a bit of renaming 3 <-> 4 here and there. 
+It worked but `coqide` crashed pretty soon when used due to this poor setup.
+
+So i decided to take the other way : port `gtksourceview3` properly.
+I achieved that last month, the PR is in phase of polishing, but was enough for me to carry on the work further.
+
+### The journey continues
+During my tasks, i encounter some annoying behabviour by `opam` :
+the `opam install` command uses the _Ocaml_ *dune* build system. But in some cases *dune* fails with a rather cryptic (at this time for me, not for more experimented Haiku developers than me) message :
+`execve() : Operation not supported`.
+At first, i thought it was because of the spawn submodule of dune. So i tried for months to do a pull request (PR) on Haiku, to kill other threads and do a fork() from another thread of the team than the main thread. I could never finish that because the main thread is very special, so you can't easily do a `fork()` from a non-main thread. But indeed it was in a file which calls execve (which is incompatible with `fork()` on Haiku if the team has more than one thread. So i disabled the call to execve(), as for the Win32 platform.
+
+### Finally, land is in sight
+When i am able to have a working `dune`, i can compile Coq : <img src="/files/blog/anarchos/coqide.png"/>
+
+### Ulysses goes home
+with coq available, i was able to compile `ccomp`, the CompCert certified compiler.
+First test : `
+[Rocker] ~/OCAML/Compcert> ccomp
+ccomp: error: no input file
+1 error detected.
+`
+At last, the binary runs.
+
+Next test : compile "int main(int argc, char**argv) {return 0;}" and runs `./a.out;echo $?` --> 0
+To be sure : compile "int main(int argc, char**argv) {return 1;}" and runs `./a.out;echo $?` --> 1
+
+Success !
+
+Now i try to use standard headers (an include to <stdio.h>:
+` ccomp test.c 
+In file included from /boot/system/develop/headers/posix/sys/types.h:11,
+                 from /boot/system/develop/headers/posix/stdio.h:9,
+                 from /boot/system/develop/headers/bsd/stdio.h:9,
+                 from test.c:1:
+/boot/system/develop/headers/os/BeBuild.h:59:9: error: #error Unsupported compiler!
+   59 | #       error Unsupported compiler!
+      |         ^~~~~
+ccomp: error: preprocessor command failed with exit code 1 (use -v to see invocation)
+`
```

Comment:
````suggestion
```
ccomp test.c 
In file included from /boot/system/develop/headers/posix/sys/types.h:11,
                 from /boot/system/develop/headers/posix/stdio.h:9,
                 from /boot/system/develop/headers/bsd/stdio.h:9,
                 from test.c:1:
/boot/system/develop/headers/os/BeBuild.h:59:9: error: #error Unsupported compiler!
   59 | #       error Unsupported compiler!
      |         ^~~~~
ccomp: error: preprocessor command failed with exit code 1 (use -v to see invocation)
```
````